pkg/controller/keyvault/ingress_tls.go

```diff
@@ -0,0 +1,104 @@
+	"strings"
+
+	"github.com/Azure/aks-app-routing-operator/pkg/config"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/controllername"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/metrics"
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	"github.com/go-logr/logr"
+	netv1 "k8s.io/api/networking/v1"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var ingressTlsControllerName = controllername.New("keyvault", "ingress", "tls")
+
+// ingressTlsReconciler manages the TLS spec of an Ingress object using the Key Vault integration. Ingresses with
+// the App Routing Key Vault annotation will have their TLS spec populated with the Key Vault secret name.
+type ingressTlsReconciler struct {
+	client         client.Client
+	events         record.EventRecorder
+	ingressManager IngressManager
+}
+
+func NewIngressTlsReconciler(manager ctrl.Manager, conf *config.Config, ingressManager IngressManager) error {
+	metrics.InitControllerMetrics(ingressTlsControllerName)
+
+	if conf.DisableKeyvault {
+		return nil
+	}
+
+	return ingressTlsControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&netv1.Ingress{}), manager.GetLogger(),
+	).Complete(&ingressTlsReconciler{
+		client:         manager.GetClient(),
+		events:         manager.GetEventRecorderFor("aks-app-routing-operator"),
+		ingressManager: ingressManager,
+	})
+}
+
+func (i *ingressTlsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(ingressTlsControllerName, res, err)
+	}()
+
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return ctrl.Result{}, errors.New("getting logger from context")
+	}
+	logger = ingressTlsControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	logger.Info("getting Ingress")
+	ing := &netv1.Ingress{}
+	if err = i.client.Get(ctx, req.NamespacedName, ing); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	managed, err := i.ingressManager.IsManaging(ing)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("checking if ingress is managed: %w", err)
+	}
+
+	if !managed {
+		logger.Info("ingress is not managed by app routing")
+		return ctrl.Result{}, nil
+	}
+
+	if val, ok := ing.Annotations[tlsCertManagedAnnotation]; !ok || strings.ToLower(val) != "true" {
+		logger.Info("ingress does not have managed annotation")
+		return ctrl.Result{}, nil
+	}
+
+	if _, ok := ing.Annotations[tlsCertKvUriAnnotation]; !ok {
+		logger.Info("ingress does not have keyvault annotation")
+		i.events.Eventf(ing, "Warning", "KeyvaultUriAnnotationMissing", "Ingress has %[1]s annotation but is missing %[2]s annotation. %[2]s annotation is needed to manage Ingress TLS.", tlsCertManagedAnnotation, tlsCertKvUriAnnotation)
+		return ctrl.Result{}, nil
+	}
+
+	logger.Info("adding TLS spec to ingress")
```

Comment:
What if a TLS definition already exists? I think we at least need some logging in here that we're overwriting existing TLS defintions

Reply:
seems fair to log it but we do need to override them.

> If the Ingress contains both required Ingresses we always set the TLS section of the Ingress to point to the App Routing managed Secret. We also update the TLS hosts to exactly match the hosts section of the rules part of the Ingress spec. We do this regardless of if the TLS section was empty before. This is a truly "managed" experience. If we chose not to reconcile if the TLS section was non-empty then we wouldn't be able to handle the case of a user adding a new Rule and hostname to the spec of the Ingress.

otherwise, we wouldn't be able to handle them just adding new hosts to the rules

Reply:
And what about rules with no hosts? Is it possible to create an ingress with a rule that has no host but has the TLS host be provided? The host replacement below would break on this case

Reply:
k8s docs say https://kubernetes.io/docs/concepts/services-networking/ingress/#tls

> hosts in the tls section need to explicitly match the host in the rules section.

So, it's not possible for an Ingress to have a no host rule but have the tls host. 

Your point is good though. I'll change the logic to ensure that host is actually defined for the rule because it's possible for it not to be for some routes.

Reply:
added log + handle rule with no host